accepted/2020/linking-libraries.md

```diff
@@ -1,170 +1,572 @@
+needs.
 
-With the addition of the Linker Reflection Flow features above, we will need to add attributes to the .NET Library methods that rely on Reflection. At a high-level, we can run the linker on the libraries and look at the warnings it produces. We then need to analyze the correct fix for the linker warning. Possibilities are:
-
-1. Add a `DynamicallyAccessedMembers` attribute to the code, to tell the linker which members need to be accessed dynamically.
-    - This attribute is similar in nature to the nullable annotations that were added in C# 8 in that they are both viral. `DynamicallyAccessedMembers` needs to be applied from the actual usage of Reflection all the way through the call tree to the location that specifies the Type, which may be a public API.
-
-2. In some cases the usage is actually safe. For example, if an "Equals" method is written by retrieving all the fields on a type and looping over them comparing each field. If a field is unused, and therefore trimmed by the linker, there is no need to compare that field for equality. In these cases we can suppress the linker warning using the `[UnconditionalSuppressMessage]` attribute.
-
-3. In some cases the API itself is not able to be linker-friendly. In these cases we don't want to suppress the warning completely, but instead pass the warning to the caller of the API. In these cases we can use the `[RequiresUnreferencedCode]` attribute. This attribute tells the linker to not warn about this method, but instead warn about any callers of this method.
-
-4. In cases where the linker is not able to infer the dependency, or the developer wants to be explicit, the `[DynamicDependency]` attribute states an explicit dependency from one method to another. When the linker sees the attributed method needs to be kept, it will also preserve anything that method has a dynamic dependency to.
-
-A high-level goal is that the .NET Libraries are clean of linker warnings. This is one level of confidence we can have that .NET applications will be linker safe.
+The trimming tool has built-in knowledge of some of the most common problematic
+patterns and relies on annotations like `RequiresUnreferencedCodeAttribute` or
+`DynamicallyAccessedMembersAttribute` to determine the rest. If it encounters
+a pattern which it can't recognize or handle for other reasons it generates
+a trim analysis warning.
+
+Such warning tells the user that there's code in the application which may
+require additional dependency which the tool could not figure out and thus
+it's possible the app will not work correctly after trimming.
+
+See the [Linker Reflection Flow Design](https://github.com/mono/linker/blob/master/docs/design/reflection-flow.md)
+for more information on how the process works.
+
+#### Annotating code
+
+Some library code can use reflection or other problematic patterns and thus
+running the trimming tool on it can break its functionality and it generates
+trim analysis warnings.
```

Comment:
```suggestion
Some library code can use reflection or other problematic patterns such that
running the trimming tool on it can break its functionality and generate
trim analysis warnings.
```